AstSemantics.md

```diff
@@ -111,6 +112,9 @@ Integer loads can specify a *storage size* which is smaller than the result type
 well as a signedness which determines whether the bytes are sign- or zero-
 extended into the result type.
 
+Each `load` and `store` specifies the [linear memory index](Modules.md#index-spaces)
+to operate on with an unsigned integer immediate.
+
```

Comment:
@JSStats I don't know what memory access performance challenge you're referring to.  If an engine wants to go the route of baking-in addresses/lengths, it can do that just as easily for multiple heaps.  If an engine wants to pin a register, it can still do that for the first (and usually only) memory.  Otherwise, the engine can hoist things into registers as needed locally.

I *can* actually see a GC `Memory`-reference type getting equivalent performance (if we stipulate things like a GC reference type that says not just `Memory`, but also "with `initial` >= `X`" etc).  I'd like wasm to be able to run with full functionality w/o requiring GC (so you could have wasm-only workers that simply don't create a GC heap or any of the JS-runtime data), but that could also be made to work when GC reference types are only used for locals (on the stack), never GC allocation.

@kripken In addition to "file system impl", "asm.js globals" is another use cases where you have a simple, static separation.  Another cool use case could be a *San tool that wants to maintain a whole shadow data structure to reflect memory; a separate memory is exactly what you want; with a single memory, you'd have to prematurely fix the heap size to split "shadow" from "real" and then "real" can also corrupt "shadow".  Over time, it's easy to imagine more use cases arising since this is an inherently artificial limitation.

Also, if you take a modular-web-app POV (instead of a monolithic-C++-app POV), there's not just 1 big heap; you have lots of modules in lots of heaps.  You don't want to force them all into the same heap (where they may clobber each other) or force them to do unnecessary copying if they form a data pipeline.  GC-`Memory` references handles the general case, but imports/exports would allow the pipeline to be declaratively composed using the plain old module system.  Again, not a killer feature, but (1) not hard (if we do GC Memories, multi-Memory imports are just a simple/special-case), (2) obvious symmetry in design given that we have multi function/table export (and probably other things too in the future); given the other things, someone would naturally ask "why *not* multiple memories?" and I don't think we have a strong concrete answer.  Again, if it's too much work as judged by engine authors, we could constrain 0..1 in MVP, but if we don't provision in the MVP, we'll have to duplicate a lot of stuff (more than just load/store ops) later which will be inelegant.